src/Apropos/Gen.hs

```diff
@@ -181,75 +221,115 @@ type GenContext = H.Gen
 
 type Generator = ExceptT GenException (WriterT (Set String) GenContext)
 
-gen :: Gen a -> Generator a
-gen (Free (Label s next)) = lift (tell (Set.singleton s)) >> gen next
-gen (Free (FailWithFootnote s _)) = throwE $ GenException s
-gen (Free (GenBool next)) = lift HGen.bool >>= (gen . next)
-gen (Free (GenInt r next)) = do
-  i <- lift $ HGen.int (hRange r)
-  gen $ next i
-gen (Free (GenList r g next)) = do
+data FreeInterleaved next where
+  G :: forall a next. Generator a -> (a -> next) -> FreeInterleaved next
+  P :: forall a next. PropertyT IO a -> (a -> next) -> FreeInterleaved next
+
+instance Functor FreeInterleaved where
+  fmap f (G g next) = G g (fmap f next)
+  fmap f (P p next) = P p (fmap f next)
+
+type Interleaved = Free FreeInterleaved
+
+mapG :: (forall a. Generator a -> Generator a) -> Interleaved b -> Interleaved b
+mapG m (Free (G g next)) = Free (G (m g) (mapG m <$> next))
+mapG m (Free (P p next)) = Free (P p (mapG m <$> next))
+mapG _ (Pure r) = Pure r
+
+gStep :: Interleaved a -> Generator (Either (Maybe (Interleaved a)) a)
+gStep (Pure a) = pure $ Right a
+gStep (Free (G g f)) = Left . Just . f <$> g
+gStep _ = pure $ Left Nothing
+
+gSteps :: Interleaved a -> Generator (Either (Interleaved a) a)
+gSteps m = do
+  mg <- gStep m
+  case mg of
+    Right a -> pure $ Right a
+    Left Nothing -> pure $ Left m
+    Left (Just so) -> gSteps so
```

Comment:
This binds together `G` computations until we either  have a result or need to run an `Interleaved` continuation that starts with a `P`.